### PR TITLE
fix(grid3d): coloring by discrete property fixed.

### PR DESCRIPTION
**Pull request by LeonidPolukhin**
For coloring by discrete property fragment shader uniform colorLookupTolerance is added.
Its value is used to define whether a property value is in range of colorMapRange and to lookup color in color table texture.

### CODE REVIEW

**Review by hkfb — APPROVED** [review]

File: typescript/packages/subsurface-viewer/src/layers/grid3d/privateGrid3dLayer.ts [hunk 1]
```diff
@@ -263,12 +264,16 @@ export default class PrivateLayer extends Layer<PrivateLayerProps> {
...
     } {
         if (this.props.colorMapFunction instanceof Uint8Array) {
+            const count = this.props.colorMapFunction.length / 3;
             return {
                 imageData: this.props.colorMapFunction,
-                count: this.props.colorMapFunction.length / 3,
+                count,
                 parameters: DISCRETE_TEXTURE_PARAMETERS,
+                //As the colors are not interpolated a slight offset in the texture is needed to avoid "color fighting" when a color is picked on the border between two colors.
+                colorLookupTolerance: (1.0 / count) * 0.5,
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/privateGrid3dLayer.ts:276]: Can count be 0 (if no/empty colormap is sent) ?

File: typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts [hunk 2]
```diff
@@ -37,11 +39,11 @@ vec4 getPropertyColor (float propertyValue) {
          // Use min/max color to clamp.
          x = max(0.0, x);
          x = min(1.0, x);
-         color = texture2D(colormap, vec2(x, 0.5));
+         color = texture2D(colormap, vec2(x + colorLookupTolerance , 0.5));
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts:42]: Is this really needed ? x is clamped to [0, 1]

File: typescript/packages/subsurface-viewer/src/layers/grid3d/privateGrid3dLayer.ts (hunk 1 above)
Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/privateGrid3dLayer.ts:294]: This case is fixed.

File: typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts (hunk 2 above)
Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts]: The shader has been reworked for discrete coloring

**Review by t0oF-azpn — APPROVED** [review]

**Pull request merged**